eth/downloader/downloader_test.go

```diff
@@ -316,9 +320,7 @@ func (dl *downloadTester) InsertReceiptChain(blocks types.Blocks, receipts []typ
 			// Migrate from active db to ancient db
 			dl.ancientHeaders[blocks[i].Hash()] = blocks[i].Header()
 			dl.ancientChainTd[blocks[i].Hash()] = new(big.Int).Add(dl.ancientChainTd[blocks[i].ParentHash()], blocks[i].Difficulty())
-
 			delete(dl.ownHeaders, blocks[i].Hash())
-			delete(dl.ownChainTd, blocks[i].Hash())
```

Comment:
Maybe it's not good to not delete the td. We can do the same thing for header by offering an internal `getTd` function and use it to retrieve from both `ownChainTd ` and `ancientChainTd`

Reply:
We could, but currently the my feeling is that the test-internal division between ancient vs own is a bit haphazard. Might be good to do a more thorough refactoring of that, which wasn't really my intention in this PR

Reply:
Seems it's trivial enough to do it

```go
diff --git a/eth/downloader/downloader_test.go b/eth/downloader/downloader_test.go
index 1b9f00c69..eb3c6fe7f 100644
--- a/eth/downloader/downloader_test.go
+++ b/eth/downloader/downloader_test.go
@@ -232,6 +232,13 @@ func (dl *downloadTester) GetTd(hash common.Hash, number uint64) *big.Int {
 	dl.lock.RLock()
 	defer dl.lock.RUnlock()
 
+	return dl.getTd(hash)
+}
+
+// getHeaderByHash retrieves the block's total difficulty if found either within
+// ancients or own blocks).
+// This method assumes that the caller holds at least the read-lock (dl.lock)
+func (dl *downloadTester) getTd(hash common.Hash) *big.Int {
 	if td := dl.ancientChainTd[hash]; td != nil {
 		return td
 	}
@@ -267,7 +274,9 @@ func (dl *downloadTester) InsertHeaderChain(headers []*types.Header, checkFreq i
 		}
 		dl.ownHashes = append(dl.ownHashes, hash)
 		dl.ownHeaders[hash] = header
-		dl.ownChainTd[hash] = new(big.Int).Add(dl.ownChainTd[header.ParentHash], header.Difficulty)
+
+		td := dl.getTd(header.ParentHash)
+		dl.ownChainTd[hash] = new(big.Int).Add(td, header.Difficulty)
 	}
 	return len(headers), nil
 }
@@ -290,7 +299,9 @@ func (dl *downloadTester) InsertChain(blocks types.Blocks) (i int, err error) {
 		dl.ownBlocks[block.Hash()] = block
 		dl.ownReceipts[block.Hash()] = make(types.Receipts, 0)
 		dl.stateDb.Put(block.Root().Bytes(), []byte{0x00})
-		dl.ownChainTd[block.Hash()] = new(big.Int).Add(dl.ownChainTd[block.ParentHash()], block.Difficulty())
+
+		td := dl.getTd(block.ParentHash())
+		dl.ownChainTd[block.Hash()] = new(big.Int).Add(td, block.Difficulty())
 	}
 	return len(blocks), nil
 }
```

But I agree that we should rework the testing a bit. It's fine to leave the old code here.

Reply:
I suggest we merge this as is now. This diff is not 100% complete, we need to add the td to ancients aswell

Reply:
Ok, committed your suggestion + added back the delete